File: src/client/nxmc/java/src/main/java/org/netxms/nxmc/modules/objects/propertypages/VNC.java
/**
 * NetXMS - open source network management system
 * Copyright (C) 2003-2024 Victor Kirhenshtein
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package org.netxms.nxmc.modules.objects.propertypages;

import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.jface.dialogs.MessageDialog;
import org.eclipse.swt.SWT;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Control;
import org.netxms.client.NXCObjectModificationData;
import org.netxms.client.NXCSession;
import org.netxms.client.objects.AbstractNode;
import org.netxms.client.objects.AbstractObject;
import org.netxms.nxmc.Registry;
import org.netxms.nxmc.base.jobs.Job;
import org.netxms.nxmc.base.widgets.LabeledText;
import org.netxms.nxmc.base.widgets.PasswordInputField;
import org.netxms.nxmc.localization.LocalizationHelper;
import org.netxms.nxmc.modules.objects.widgets.ObjectSelector;
import org.netxms.nxmc.tools.WidgetHelper;
import org.xnap.commons.i18n.I18n;

/**
 * "VNC" property page for node
 */
public class VNC extends ObjectPropertyPage
{
   private I18n i18n = LocalizationHelper.getI18n(VNC.class);

   private AbstractNode node;
   private ObjectSelector vncProxy;
   private LabeledText vncPort;
   private PasswordInputField vncPassword;

   /**
    * Create new page.
    *
    * @param object object to edit
    */
   public VNC(AbstractObject object)
   {
      super(LocalizationHelper.getI18n(VNC.class).tr("VNC"), object);
   }

   /**
    * @see org.netxms.nxmc.modules.objects.propertypages.ObjectPropertyPage#getId()
    */
   @Override
   public String getId()
   {
      return "communication.vnc";
   }

   /**
    * @see org.netxms.nxmc.modules.objects.propertypages.ObjectPropertyPage#getParentId()
    */
   @Override
   public String getParentId()
   {
      return "communication";
   }

   /**
    * @see org.netxms.nxmc.modules.objects.propertypages.ObjectPropertyPage#isVisible()
    */
   @Override
   public boolean isVisible()
   {
      return object instanceof AbstractNode;
   }

   /**
    * @see org.eclipse.jface.preference.PreferencePage#createContents(org.eclipse.swt.widgets.Composite)
    */
   @Override
   protected Control createContents(Composite parent)
   {
      node = (AbstractNode)object;

      Composite dialogArea = new Composite(parent, SWT.NONE);
      GridLayout dialogLayout = new GridLayout();
      dialogLayout.marginWidth = 0;
      dialogLayout.marginHeight = 0;
      dialogLayout.numColumns = 2;
      dialogArea.setLayout(dialogLayout);

      vncPort = new LabeledText(dialogArea, SWT.NONE);
      vncPort.setLabel("Port");
      GridData gd = new GridData(SWT.FILL, SWT.CENTER, false, false);
      gd.widthHint = WidgetHelper.getTextWidth(vncPort, "00000") * 2;
      vncPort.setLayoutData(gd);
      vncPort.setText(Integer.toString(node.getVncPort()));
      vncPort.getTextControl().setTextLimit(5);

      vncPassword = new PasswordInputField(dialogArea, SWT.NONE);
      vncPassword.setLabel(i18n.tr("Password"));
      vncPassword.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, true, false));
      vncPassword.setText(node.getVncPassword());

      vncProxy = new ObjectSelector(dialogArea, SWT.NONE, true);
      vncProxy.setLabel(i18n.tr("Proxy"));
      vncProxy.setEmptySelectionName("<default>");
      vncProxy.setObjectId(node.getVncProxyId());
      vncProxy.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, true, false, 2, 1));

      return dialogArea;
   }

   /**
    * Apply changes
    * 
    * @param isApply true if update operation caused by "Apply" button
    */
   protected boolean applyChanges(final boolean isApply)
   {
      final NXCObjectModificationData md = new NXCObjectModificationData(node.getObjectId());

      if (isApply)
         setValid(false);

      try
      {
         md.setVncPort(Integer.parseInt(vncPort.getText(), 10));
      }
      catch(NumberFormatException e)
      {
         MessageDialog.openWarning(getShell(), i18n.tr("Warning"), i18n.tr("Please enter valid VNC port number"));
         if (isApply)
            setValid(true);
         return false;
      }

      md.setVncProxy(vncProxy.getObjectId());
      md.setVncPassword(vncPassword.getText());

      final NXCSession session = Registry.getSession();
      new Job(i18n.tr("Updating VNC settings for node {0}", node.getObjectName()), null, messageArea) {
         @Override
         protected void run(IProgressMonitor monitor) throws Exception
         {
            session.modifyObject(md);
         }

         @Override
         protected String getErrorMessage()
         {
            return i18n.tr("Cannot update VNC settings for node {0}", node.getObjectName());
         }

         @Override
         protected void jobFinalize()
         {
            if (isApply)
               runInUIThread(() -> VNC.this.setValid(true));
         }
      }.start();
      return true;
   }

   /**
    * @see org.eclipse.jface.preference.PreferencePage#performDefaults()
    */
   @Override
   protected void performDefaults()
   {
      super.performDefaults();
      vncProxy.setObjectId(0);
      vncPassword.setText("");
      vncPort.setText("5900");
   }
}

File: src/server/core/vnc.cpp
/*
** NetXMS - Network Management System
** Copyright (C) 2003-2025 Raden Solutions
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
**
** File: vnc.cpp
**
**/

#include "nxcore.h"

/**
 * Check VNC connection using given communication settings
 */
bool VNCCheckConnection(Node *proxyNode, const InetAddress& addr, uint16_t port)
{
   StringBuffer request(_T("VNC.ServerState("));
   TCHAR ipAddr[64];
   request
      .append(addr.toString(ipAddr))
      .append(_T(','))
      .append(port)
      .append(_T(')'));

   TCHAR response[2];
   return proxyNode->getMetricFromAgent(request, response, 2) == DCE_SUCCESS && response[0] == _T('1');
}

/**
 * Check VNC connection using given communication settings
 */
bool VNCCheckConnection(uint32_t proxyNodeId, const InetAddress& addr, uint16_t port)
{
   shared_ptr<NetObj> proxyNode = FindObjectById(proxyNodeId, OBJECT_NODE);
   return (proxyNode != nullptr) ? VNCCheckConnection(static_cast<Node*>(proxyNode.get()), addr, port) : false;
}

/**
 * Determine VNC communication settings for node
 * On success, returns true and fills selectedPort
 */
bool VNCCheckCommSettings(uint32_t proxyNodeId, const InetAddress& addr, int32_t zoneUIN, uint16_t *selectedPort)
{
   TCHAR ipAddrText[64];
   addr.toString(ipAddrText);

   shared_ptr<NetObj> proxyNode = FindObjectById(proxyNodeId, OBJECT_NODE);
   if (proxyNode == nullptr)
   {
      nxlog_debug_tag(DEBUG_TAG_VNC, 5, _T("VNCCheckCommSettings(%s): invalid proxy node ID %u"), ipAddrText, proxyNodeId);
      return false;
   }

   IntegerArray<uint16_t> ports = GetWellKnownPorts(_T("vnc"), zoneUIN);

   bool success = false;
   for (int i = 0; i < ports.size(); i++)
   {
      uint16_t port = ports.get(i);
      nxlog_debug_tag(DEBUG_TAG_VNC, 5, _T("VNCCheckCommSettings(%s): trying port %d"), ipAddrText, port);
      success = VNCCheckConnection(static_cast<Node*>(proxyNode.get()), addr, port);
      if (success)
      {
         if (selectedPort != nullptr)
            *selectedPort = port;
         break;
      }
   }

   nxlog_debug_tag(DEBUG_TAG_VNC, 5, _T("VNCCheckCommSettings(%s): %s"), ipAddrText, success ? _T("success") : _T("failure"));
   return success;
}
